src/Index.hpp

```diff
@@ -110,6 +110,131 @@ namespace nbs {
+                    }
+                    else {
+                        throw std::invalid_argument("No packet found for this type at the requested timestamp");
+                    }
+
+                    states.push_back(TimestampState{0, position, begin, end});
+                }
+            }
+
+            if (states.empty()) {
+                throw std::runtime_error("File with requested types exist but contain no messages. ");
+            }
+
+            // Traverse through input type(s)
+            while (true) {
+                if (steps > 0) {
+                    auto min_val = std::min_element(  //
+                        states.begin(),
+                        states.end(),
+                        [](TimestampState& a, TimestampState& b) {
+                            auto a_time = std::next(a.position) == a.end ? std::numeric_limits<uint64_t>::max()
+                                                                         : std::next(a.position)->item.timestamp;
+                            auto b_time = std::next(b.position) == b.end ? std::numeric_limits<uint64_t>::max()
+                                                                         : std::next(b.position)->item.timestamp;
+                            return a_time < b_time;
+                        });
+
+                    // Check if the next step would move us off the end of the nbs file.
+                    if (std::distance(min_val->position, min_val->end) == 0) {
+                        return min_val->position->item.timestamp;
+                    }
+
+                    if (types.size() > 1) {
+                        // Check if this next step would exceed our number of steps.
+                        if (min_val->jumps + 1 == steps) {
+                            auto best =
+                                std::max_element(states.begin(),
+                                                 states.end(),
+                                                 [](const TimestampState& a, const TimestampState& b) {
+                                                     return a.position->item.timestamp < b.position->item.timestamp;
+                                                 });
+                            return best->position->item.timestamp;
+                        }
+                    }
+
+                    else {
+                        if (min_val->jumps == steps) {
+                            auto best =
+                                std::max_element(states.begin(),
+                                                 states.end(),
+                                                 [](const TimestampState& a, const TimestampState& b) {
+                                                     return a.position->item.timestamp < b.position->item.timestamp;
+                                                 });
+                            return best->position->item.timestamp;
+                        }
+                    }
```

Comment:
I think the one that's checking + 1 is incorrect as that's checking if moving one more time would get you equal to the number of steps
```suggestion
                    if (min_val->jumps == steps) {
                        auto best =
                            std::max_element(states.begin(),
                                             states.end(),
                                             [](const TimestampState& a, const TimestampState& b) {
                                                 return a.position->item.timestamp < b.position->item.timestamp;
                                             });
                        return best->position->item.timestamp;
                    }
```